Remove assumed minimum length of height in coinbase

rnicoll opened the pull request:
Cut down version of #1117 with test case added.

schildbach commented on core/src/test/java/org/bitcoinj/core/BlockTest.java:
```diff
@@ -166,4 +167,20 @@ public void testUpdateLength() {
+        // contains a coinbase transaction whose height is two bytes, which is
+        // shorter than we see in most other cases.
+
+        byte[] blockAsBytes = CoinbaseBlockTest.getBytes(getClass().getResourceAsStream("block_testnet21066.dat"));
```
Can you use Guava here?

```
Block block = params.getDefaultSerializer().makeBlock(
    ByteStreams.toByteArray(getClass().getResourceAsStream("block.dat")));
```

kparmar1 commented:
This will fail later on (at least on testnet).... I'll give you the block number when I can.

kparmar1 commented on core/src/main/java/org/bitcoinj/script/ScriptBuilder.java:
```diff
@@ -351,10 +351,17 @@ public static Script createOpReturnScript(byte[] data) {
+
+        // Strip any zeroes from the end
+        int size;
+        for (size = 0; size < int32Buffer.length; size++) {
```
You can't do this. To strip zeros you have to go from the back (it's little endian format).

kparmar1 replied:
Says the little Indian :-)

schildbach commented on core/src/test/java/org/bitcoinj/core/BlockTest.java:
```diff
@@ -166,4 +168,19 @@ public void testUpdateLength() {
+        // contains a coinbase transaction whose height is two bytes, which is
+        // shorter than we see in most other cases.
+
+        Block block = params.getDefaultSerializer().makeBlock(
```
Sorry, those are the wrong params. `TestNet3Params.get()` should be right.

rnicoll commented:
Corrected testnet params, and fixed a bug in the block height trimming code that meant it would stop at the first zero, irrespective of whether there was data after. It now trims from the right as suggested by @kparmar1 .

Add unit tests covering the script height generation too.

schildbach commented on core/src/test/java/org/bitcoinj/core/BlockTest.java:
```diff
@@ -26,6 +28,7 @@
 import java.math.BigInteger;
 import java.util.Arrays;
 import java.util.EnumSet;
+import static org.bitcoinj.core.CoinbaseBlockTest.params;
```
This import can be removed.

schildbach commented on core/src/test/java/org/bitcoinj/script/ScriptTest.java:
```diff
@@ -412,4 +412,24 @@ public void getToAddress() throws Exception {
     public void getToAddressNoPubKey() throws Exception {
         ScriptBuilder.createOutputScript(new ECKey()).getToAddress(params, false);
     }
+
```
Great, thanks for this unit test!

schildbach commented:
Maybe mention Kalpesh Parmar in the commit comment for finding the issue?

rnicoll commented:
Good idea - done, and added a note explaining the change in the code, with a credit there too.

schildbach commented on core/src/test/java/org/bitcoinj/core/BlockTest.java:
```diff
@@ -166,4 +168,19 @@ public void testUpdateLength() {
+            ByteStreams.toByteArray(getClass().getResourceAsStream("block_testnet21066.dat")));
+
+        // Check block.
+        assertNotNull(block);
```
Hmm, I think this assert is useless.

kparmar1 commented:
No that's fine - you don't have to mention me (I'm glad to help when I can) :-)

schildbach commented on core/src/main/java/org/bitcoinj/script/ScriptBuilder.java:
```diff
@@ -119,6 +123,65 @@ public ScriptBuilder smallNum(int index, int num) {
+            //    0x80 to it, since it will be subtracted and interpreted as a negative when
+            //    converting to an integral.
+
+            if ((result.peek().byteValue() & 0x80) != 0)
```
Parenthesis missing.

schildbach commented on core/src/main/java/org/bitcoinj/script/ScriptBuilder.java:
```diff
@@ -119,6 +123,65 @@ public ScriptBuilder smallNum(int index, int num) {
+                absvalue >>= 8;
+            }
+
+            //    - If the most significant byte is >= 0x80 and the value is positive, push a
```
Why 4 spaces between "//" and the comment?

rnicoll replied:
Mostly because those notes are copied directly from the C++ code. Cleaning them up now and will re-push.

schildbach commented on core/src/main/java/org/bitcoinj/script/ScriptBuilder.java:
```diff
@@ -119,6 +123,65 @@ public ScriptBuilder smallNum(int index, int num) {
+            long absvalue = Math.abs(num);
+
+            while (absvalue != 0)
+            {
```
Parenthesis should go to end of previous line.

schildbach commented:
Thanks for this work! I annotated some minor style things.

@kparmar1 Do you think you could test this patch with whatever you were doing and see if it works?

kparmar1 commented:
Yes I'll do that tomorrow - NP.

luke-jr commented:
FYI: The BIP34 height is always and required to be minimum length. So I suspect this fix is wrong.

schildbach commented:
@luke-jr Where do you read that from? BIP34 says:

> The format of the height is "serialized CScript" -- first byte is number of bytes in the number (will be 0x03 on main net for the next 300 or so years), following bytes are little-endian representation of the number.

rnicoll commented:
It is minimum length, as @luke-jr suggested elsewhere, I'd forgotten the sign bit. Specifically, I'd forgotten if the most significant byte had the 0x80 bit set and the value was not negative, a zero is appended to indicate the value is positive.

luke-jr commented:
@schildbach Core source code, which is the only *specification* of the Bitcoin protocol.

rnicoll commented:
Clean up done and pushed

schildbach commented on core/src/main/java/org/bitcoinj/script/ScriptBuilder.java:
```diff
@@ -119,6 +123,53 @@ public ScriptBuilder smallNum(int index, int num) {
+
+        // This encoding is copied directly from
+        // https://github.com/bitcoin/bitcoin/blob/b48da5c1894a70f8fa2a50deb2e056c38ed27ecb/src/script/script.h#L303
+        // Specifically note that the encoded format is not always the minimum
```
Hmm is this statement still true?

schildbach commented on core/src/test/java/org/bitcoinj/script/ScriptTest.java:
```diff
@@ -412,4 +412,38 @@ public void getToAddress() throws Exception {
+            0x01,        // Length of the pushed data
+            ((byte) 133) // Pushed data
+        }, builder.build().getProgram());
+    }
```
A nice edge case for an additional test would be bigNum(0).

rnicoll commented:
Comments updated. Although I don't think this applies to either main or test network, strictly low heights (0-16) are encoded as op-codes, so have updated this to use small/big number encoding as appropriate.

schildbach commented:
Great, I'll merge this soon. @kparmar1 Have you been able to test this patch?

kparmar1 commented:
I'm syning testnet at the moment....so far so good but I'll leave it running over night to complete,but hopefully it will finish before I go to bed.

Sent from TypeMail



On 22 Nov 2015 19:39, at 19:39, Andreas Schildbach <notifications@github.com> wrote:
>Great, I'll merge this soon. @kparmar1 Have you been able to test this
>patch?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoinj/bitcoinj/pull/1119#issuecomment-158792742

kparmar1 commented:
This is running fine (75% synced so far- now it's very slow to complete).  I'll report back when it does.

FYI:
I did see a couple (half dozen maybe) warnings, never seen before. This is def not related to the height though:-

19:41:03.261 [Script verification] WARN  org.bitcoinj.script.Script - Signature checking failed!
java.lang.RuntimeException: java.io.EOFException: EOF found when length expected
	at org.bitcoinj.core.ECKey$ECDSASignature.decodeFromDER(ECKey.java:591) ~[bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at org.bitcoinj.crypto.TransactionSignature.decodeFromBitcoin(TransactionSignature.java:188) ~[bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at org.bitcoinj.script.Script.executeCheckSig(Script.java:1323) [bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at org.bitcoinj.script.Script.executeScript(Script.java:1265) [bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at org.bitcoinj.script.Script.correctlySpends(Script.java:1502) [bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at org.bitcoinj.core.FullPrunedBlockChain$Verifier.call(FullPrunedBlockChain.java:171) [bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at org.bitcoinj.core.FullPrunedBlockChain$Verifier.call(FullPrunedBlockChain.java:154) [bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_51]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_51]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_51]
	at org.bitcoinj.utils.ContextPropagatingThreadFactory$1.run(ContextPropagatingThreadFactory.java:35) [bitcoinj-core-0.14-SNAPSHOT.jar:na]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_51]
Caused by: java.io.EOFException: EOF found when length expected
	at org.spongycastle.asn1.ASN1InputStream.readLength(ASN1InputStream.java:332) ~[core-1.51.0.0.jar:na]
	at org.spongycastle.asn1.ASN1InputStream.readLength(ASN1InputStream.java:112) ~[core-1.51.0.0.jar:na]
	at org.spongycastle.asn1.ASN1InputStream.readObject(ASN1InputStream.java:237) ~[core-1.51.0.0.jar:na]
	at org.bitcoinj.core.ECKey$ECDSASignature.decodeFromDER(ECKey.java:577) ~[bitcoinj-core-0.14-SNAPSHOT.jar:na]
	... 11 common frames omitted

kparmar1 commented:
LGTM

Fully synced up on testnet with the PR (nice one, the patch works:-)

schildbach commented:
Merged. Many thanks for finding and fixing this!

schildbach commented:
Looks like height in [coinbase is signed](https://github.com/bitcoin/bips/pull/318) -- is that relevant our changes?

rnicoll commented:
Just checked the code, should all be fine. We encode the height via the normal script data encoder, which handles negatives by default. Might have been a problem if it didn't have a signed bit, as that would differ to other numbers, though.

schildbach commented:
Thanks for checking (-:
